src/plugins/keras_support/categorical_crossentropy_operation.cpp

```diff
@@ -0,0 +1,291 @@
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include <type_traits>
+
+#include <blaze/Math.h>
+#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
+#include <blaze_tensor/Math.h>
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const cat_cross_operation::match_data =
+    {
+        hpx::util::make_tuple("categorical_crossentropy",
+        std::vector<std::string>{
+            "categorical_crossentropy(_1,_2)",
+            "categorical_crossentropy(_1,_2,_3)"
+        },
+        &create_cat_cross_operation, &create_primitive<cat_cross_operation>,
+        R"(target, output, from_logits
+        Args:
+
+            target (array_like) : input array
+            output (array_like) : output array
+            from_logits (boolean): boolean value
+
+        Returns:
+            The value should be the same as would be returned by the following
+            Python function:
+
+            def cat_cross(target, output, from_logits=False):
+                if from_logits:
+                    output = softmax(output)
+                else:
+                    output /= output.sum(axis=-1, keepdims=True)
+                output = np.clip(output, 1e-7, 1 - 1e-7)
+                return np.sum(target * -np.log(output), axis=-1, keepdims=False)
+        )")
+    };
+    const double clip_low = 1e-7;
+    const double clip_high = 1 - clip_low;
+
+    ///////////////////////////////////////////////////////////////////////////
+    cat_cross_operation::cat_cross_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    primitive_argument_type cat_cross_operation::cat_cross0d(
+        arg_type&& target,arg_type&& output,bool from_logits) const
+    {
+        double v = 1;
+
+        if(!from_logits)
+            v = output.scalar()/output.scalar();
+
+        return primitive_argument_type{
+            static_cast<double>(-target.scalar()*blaze::log(clip_high))};
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    using matrix_type =
+         blaze::DynamicMatrix<double>;
+
+    using vector_type =
+         blaze::DynamicVector<double>;
+
+    using tensor_type =
+         blaze::DynamicTensor<double>;
+
+    ///////////////////////////////////////////////////////////////////////////
+    primitive_argument_type cat_cross_operation::cat_cross1d(
+        arg_type&& target,arg_type&& output,bool from_logits) const
+    {
+        vector_type&& output_ = output.vector();
+        vector_type&& target_ = target.vector();
+        if(from_logits)
+        {
+            output_ = blaze::softmax(output_);
```

Comment:
This always allocates new memory. In case if `output` is not a reference (i.e. holds the actual data) we could optimize things and reuse the memory owned by `output`. Please see below as this can be achieved.